src/utils/sample_utils.py

```diff
@@ -40,6 +41,53 @@ def __setattr__(self, key, value):
 
 Constraints = Bunch
 
+def predictive_probability(M_c, X_L, X_D, Y, Q):
+    # Evaluates the joint logpdf of crosscat columns. This is acheived by
+    # invoking column_value_probability on univariate columns with
+    # cascading the constraints (the chain rule).
+
+    # Q (query): list of three element tuples where each tuple, (r,c,x)
+    #  contains a row r; column, c; value x. All rows must be the same.
+    # Y (contraints), follows an identical format.
+
+    # The current interface does not allow the query columns to have different
+    # row numbers, so this function will ensure the same constraint, pending
+    # a formalization of the semantic meaning of predictive_probability of
+    # arbitrary patterns of cells.
+    seen = set()
+    for (row, col, val) in Q:
+        if col in seen:
+            raise ValueError('Cannot specify duplicate columns.')
+        if row != Q[0][0]:
+            raise ValueError('Cannot specify different rows.')
+        if len(M_c['column_metadata']) <= col:
+            raise ValueError('Cannot specify hypothetical column.')
+        seen.add(col)
+    # Ensure consistency for nodes in both query and constraints.
+    # This behavior is correct, even for real-valued datatypes. Conditional
+    # probability is itself a complex topic, but consider random
+    # variable X continuous. Then the conditional density of X f(s|X=t) is
+    # 1 if s==t and 0 otherwise. Note change of the dominating measure from
+    # Lebesgue to counting. The argument is not rigorous but correct.
+    ignore = set()
+    for (rq, cq, vq), (rv, cy, vy) in itertools.product(Q, Y):
+        if cq == cy and rq == rv:
+            if vq == vy:
+                ignore.add(cq)
+            else:
+                return -float('inf')
+    Y_prime = copy.deepcopy(Y)
+    # Chain rule.
+    prob = 0
+    for query in Q:
+        if query[1] in ignore:
+            continue
+        r = simple_predictive_probability(M_c, X_L, X_D, Y_prime, [query])
```

Comment:
Force the numpy array that comes out of this mess into a float (since you know it will have size 1 because of one query at a time), and add an assertion in the test that the type returned by predictive_probability is correct.